### PR TITLE
Parse meta[name=go-import] tag for not-well-known sites

### PR DESCRIPTION
**Pull request by hnakamur**
This pull request adds support for dynamic sites like "google.golang.org/grpc"

### CODE REVIEW

File: submodule.go [hunk 1]
```diff
@@ -68,6 +80,55 @@ func addVendorSubmodule(importpath string) error {
...
+	"hub.jazz.net/git/",
+	"git.apache.org/",
+	"git.openstack.org/",
+}
+
+func getHttpsURLForImportPath(importpath string) (url string, err error) {
+	url = "https://" + importpath
+	for _, site := range wellKnownSites {
+		if strings.HasPrefix(importpath, site) {
+			return
```
Comment (kovetskiy) [inline submodule.go]: Please do not use naked return, I think it will be more clear if we will use it as `return url, nil`

File: submodule.go [hunk 2]
```diff
@@ -68,6 +80,55 @@ func addVendorSubmodule(importpath string) error {
...
+	}
+
+	// NOTE: Parse <meta name="go-import" content="import-prefix vcs repo-root">
+	// For detail, see the output of "go help importpath"
+	var doc *goquery.Document
+	doc, err = goquery.NewDocument(url)
+	if err != nil {
+		return
+	}
+	doc.Find("meta[name=go-import]").Each(func(i int, s *goquery.Selection) {
```
Comment (kovetskiy) [inline submodule.go]: please rename `s` to `selection`, `s` it's too ambigious

File: submodule.go [hunk 3]
```diff
@@ -68,6 +80,55 @@ func addVendorSubmodule(importpath string) error {
 	return errors.New(strings.Join(errs, "\n"))
 }
 
+// NOTE: This list is copied from
+// https://github.com/golang/go/blob/10538a8f9e2e718a47633ac5a6e90415a2c3f5f1/src/cmd/go/vcs.go#L821-L861
+var wellKnownSites = []string{
```
Comment (kovetskiy) [inline submodule.go]: please move this statement before the functions

**Comment by kovetskiy** [comment]
@hnakamur thanks for the pull-request, it's must have feature for manul, I don't know how I miss it :+1: 

Btw, I think such important part of code should be tested, can you add some integration tests https://github.com/kovetskiy/manul/tree/master/tests ?

**Comment by hnakamur** [comment]
Thanks for your comment. I fixed points you commented.

As for integration test, I looked at existing tests, but I don't know how to do test for a site which has a `meta name="go-import"` tag.
If I use a real site like google.golang.org/grpc, the commit hash will change so we cannot use a fixed value.

Oh, I come up on a idea. How about setting up a project github page for github.com/kovetskiy/manul-test-foo?
The url will be https://kovetskiy.github.io/manul-test-foo and we can put a `meta name="go-import" content="kovetskiy.github.io/manul-test-foo git github.com/kovetskiy/manul-test-foo"` tag in https://kovetskiy.github.io/manul-test-foo/index.html.

I am not absolutely sure yet but I think probably it works. First I'll create a test project and try with it.
I work on weekdays so it may take some time.
Feel free to try it yourself if you don't want to wait for me.

**Comment by seletskiy** [comment]
@hnakamur: I guess, you can use [blankd](https://github.com/kovetskiy/blankd), and use package name such as `localhost.localdomain/a/b`.

**Comment by hnakamur** [comment]
@seletskiy Thanks for your comment. We need a https server listening at the standard https port 443.
blankd does not seem to support https.
Also we have to use the root user if we run a https server at port 443.
I think we should not force manul developers to run tests with the root user.

Then I tried a github page but it turned out that it does not work well for `go get` sending a request with a query parameter `?go-get=1`.

So I created an application on the Google App Engine.
The source code is at https://github.com/hnakamur/google-app-engine-manul-test-foo.

And the commit https://github.com/kovetskiy/manul/pull/18/commits/1378ed5ebcd368a51b92b6b1f7a6c0908460e897 uses this application.

I'll keep this application running within the free tier.
If it costs money in the future, I'll delete it.

Could you tell me what do you think?

**Comment by kovetskiy** [comment]
@hnakamur I think applications on Google App Engine it's too much for tests.

I think we can learn **blankd** to serving https connections (it will be very easy) and specify import path with non-standard port, something like `:61000`.

Thank you for your patience!

**Comment by hnakamur** [comment]
I tried to specify import path with non-standard port, but it does not compile.

```
$ cat main.go
package main

import foo "localhost:61000/a/b"

func main() {
        foo.Foo()
}
$ go build
can't load package: package github.com/hnakamur/tmp: 
main.go:3:12: invalid import path: "localhost:61000/a/b"
$ go version
go version go1.7rc3 linux/amd64
```

I saw [git - How Can i specify ports using golang Remote import paths? - Stack Overflow](http://stackoverflow.com/questions/34922929/how-can-i-specify-ports-using-golang-remote-import-paths) and it does not seem we can specify import path with non-standard port.

As for bankd to serving http connections, is there an example?

**Comment by kovetskiy** [comment]
@hnakamur I found some way:

```
diff --git a/imports.go b/imports.go
index 0468906..ba33425 100644
--- a/imports.go
+++ b/imports.go
@@ -99,6 +99,10 @@ func parseImports(recursive bool) ([]string, error) {
                continue
            }

+           if inTests {
+               importpath = strings.Replace(importpath, "__blankd__", "localhost:60001", -1)
+           }
+
            if isOwnPackage(importpath, cwd) {
                continue
            }
diff --git a/main.go b/main.go
index 22c3429..30d5994 100644
--- a/main.go
+++ b/main.go
@@ -1,8 +1,11 @@
 package main

 import (
+   "crypto/tls"
    "fmt"
    "log"
+   "net/http"
+   "os"
    "strconv"
    "strings"

@@ -46,8 +49,24 @@ Options:

 var (
    verbose bool
+   inTests bool
 )

+func init() {
+   for i, arg := range os.Args {
+       if arg == "--integration-test" {
+           inTests = true
+           os.Args = append(os.Args[:i], os.Args[i+1:]...)
+       }
+   }
+
+   http.DefaultClient.Transport = &http.Transport{
+       TLSClientConfig: &tls.Config{
+           InsecureSkipVerify: true,
+       },
+   }
+}
+
 func main() {
    args := godocs.MustParse(usage, version)

diff --git a/tests/https.test.sh b/tests/https.test.sh
new file mode 100644
index 0000000..a594463
--- /dev/null
+++ b/tests/https.test.sh
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+:project "main.go" <<GO
+package main
+
+import "__blankd__/kovetskiy/manul-test-foo"
+
+func main() {
+    bar.A()
+}
+GO
+
+
+tests:put server <<SRV
+#!/bin/bash
+
+cat <<HTTP
+200 OK
+
+<meta name="go-import" content="blah" />
+HTTP
+SRV
+tests:ensure chmod +x $(tests:get-tmp-dir)/server
+
+
+:lib "github.com/kovetskiy/manul-test-foo"
+tests:ensure  mv \
+    $(tests:get-tmp-dir)/go/src/github.com \
+    $(tests:get-tmp-dir)/go/src/__blankd__
+
+tests:ensure blankd \
+    -l localhost:60001 \
+    -e $(tests:get-tmp-dir)/server \
+    -o $(tests:get-tmp-dir)/blankd.log \
+    --tls
+
+tests:ensure :manul --integration-test -I localhost:60001/kovetskiy/manul-test-foo
diff --git a/tests/teardown.sh b/tests/teardown.sh
index e69de29..df260ce 100644
--- a/tests/teardown.sh
+++ b/tests/teardown.sh
@@ -0,0 +1,3 @@
+if [[ -f $(tests:get-tmp-dir)/blankd.log ]]; then
+    cat $(tests:get-tmp-dir)/blankd.log
+fi
```

**Comment by hnakamur** [comment]
@kovetskiy Thanks for your solution.
Should I include the diff for imports.go and main.go of manul in this pull request or should I create another pull request? Or could you update manul first?

**Comment by kovetskiy** [comment]
@hnakamur better to apply this changeset to your branch and finish the testing (we need check that manul visits url fron `content` attribute)

**Comment by hnakamur** [comment]
@kovetskiy OK, I'll put your change in this pull request. But I think it's better to add an option `--insecure-skip-verify` and change `init()` to

```
func init() {
   for i, arg := range os.Args {
       if arg == "--integration-test" {
           inTests = true
           os.Args = append(os.Args[:i], os.Args[i+1:]...)
       } else if arg == "--insecure-skip-verify" {
           http.DefaultClient.Transport = &http.Transport{
               TLSClientConfig: &tls.Config{
                   InsecureSkipVerify: true,
               },
           }
       }
   }
}
```

Or just change `init()` to

```
func init() {
   for i, arg := range os.Args {
       if arg == "--integration-test" {
           inTests = true
           os.Args = append(os.Args[:i], os.Args[i+1:]...)
           http.DefaultClient.Transport = &http.Transport{
               TLSClientConfig: &tls.Config{
                   InsecureSkipVerify: true,
               },
           }
       }
   }
}
```

I mean that we should only use `InsecureSkipVerify: true` for integration tests.
Could you tell me what do you think?

**Comment by kovetskiy** [comment]
@hnakamur first case looks nice, ok.

**Comment by hnakamur** [comment]
@kovetskiy OK, I've finally made an integration test for parsing meta go-import tag using blankd.
Could you take another look?
Thanks!

File: submodule.go [hunk 4]
```diff
@@ -68,6 +90,48 @@ func addVendorSubmodule(importpath string) error {
...
+			return
+		}
+		content, exists := selection.Attr("content")
+		if !exists {
+			err = fmt.Errorf(`"content" attribute not found in meta name="go-import" at %s`, url)
+			return
+		}
+		terms := strings.Fields(content)
+		if len(terms) != 3 {
+			err = fmt.Errorf(`invalid formatted "content" attribute in meta name="go-import" at %s`, url)
```
Comment (kovetskiy) [inline submodule.go]: please format lines in 80 characters per line style

File: tests/-Q-parse-meta-go-import.test.sh [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+tests:ensure chmod +x $(tests:get-tmp-dir)/server
+
+
+:lib "github.com/kovetskiy/blankd"
+:lib "github.com/kovetskiy/manul-test-foo"
+tests:ensure  mv \
+    $(tests:get-tmp-dir)/go/src/github.com \
+    $(tests:get-tmp-dir)/go/src/__blankd__
+
+tests:ensure $(tests:get-tmp-dir)/go/bin/blankd \
```
Comment (kovetskiy) [inline tests/-Q-parse-meta-go-import.test.sh:31]: wow, nice idea!

**Comment by kovetskiy** [comment]
@hnakamur merged, thanks for the contribution and patience!

**Comment by hnakamur** [comment]
@kovetskiy Thanks for your advices and reviews!

**Pull request merged**